README.md
![logo](https://i.imgur.com/YAME6yT.png "#FrostyBot")

## Summary

FrostyBot is a minimal endpoint that is designed to be used with webhooks in Tradingview alerts. It is designed to work cryptocurrency exchanges, specifically Bitmex, Deribit, FTX and Binance Futures (more will be added in future). It receives simple commands and translates them to specific exchange orders and sends them to the exchange. This allows you to write your strategies and do your backtesting on Tradingview (or python, or whatever you like), and then easily integrate your strategies with your exchange once you are happy with the backtest. This then allows you to use a Tradingview strategy or study as your trading engine, with Frostybot merely facilitating the integration to your exchange. It also allows you to integrate your exchange with any other product that can send webhooks, such as Telegram or Discord.

The way it works is simple:

* There are four main trading commands, namely **long**, **short**, **close** and **cancel**. Depending on the parameters you give these commands, you can open long or short positions, close positions, take profit (partial close) or cancel existing unfilled orders. The bot supports both limit and market orders and the commands are triggered either via a Tradingview alert webhook, or manually by you via the commandline.
* The size of the order is given using the size=xxx parameter, and can either be given in USD such as size=10000. or by a factor of your account balance, such as size=2x or size=200%
* If you specify the price using the price=xxx parameter, a limit order will be used. If you omit the price, a market order will be used at the current price.
* Going short if you are already long will place a short order for the size of your existing long position plus the amount of contracts you requested short, leaving you short the size requested. For example, if you are already long 10000 contracts and you enter a short order for 5000 contracts, the bot will actually place a sell order for 15000 contracts, leaving you 5000 contracts short after the order is filled (vice versa for short to long flips). This is done to simplify the bot logic to be written in Pinescript.
* If you are long 10000 contracts, and you enter a 15000 long order, the bot will place a buy order for only 5000 contracts, leaving you 15000 contracts long when the order is filled. If you are already long more contracts than requested, no additional order will be placed (vice versa for short orders). This allows you to control the maximum size you would like to be long or short, irrelevent of how many times Pinescript triggers the order.
* Anything that can trigger an alert in Tradingview can be used with this bot for trade execution, such as the crossing of a trendline, EMA crosses, etc. If you create your own Pinescript studies, use the alertcondition function to define alerts for buy, sell, stop loss and take profit orders as you need, and then create alerts on those alert conditions to trigger the bot. The possibilities are endless!

## Authors

Developers listed below can normally be found on the [FrostyBot Discord Server](https://discord.gg/yK4U93s). They are also known to frequent #the-lab channel on [Krown's Crypto Cave Discord Server](https://discordapp.com/invite/hzKU7qe):

* FrostyAF
* Barnz
* SemiQuasi

We kindly ask that you fully read the documentation before requesting any support, as most of the questions we get asked are already well documented.

## Dedication

Dedicated to @christiaan's mom, an all-round classy lady.

## Disclaimer
Use this bot at your own risk. The authors accept no responsibility for losses incurred through using this software. This is a 0.x release which means it's beta software. So it may and probably will have some bugs. We strongly advise you to use a sub-account with a limited balance, or a testnet account to ensure that the bot is profitable before going live with any strategy. While we have gone to great lengths to test the software, if you do find any bugs, please report them to us in the [FrostyBot Discord Server](https://discord.gg/yK4U93s) or on Github, and we will sort them out. Remember that risk management is your responsibility. If you lose your account, that's entirely on you.

## Supported Exchanges
Currently Bitmex, Deribit and FTX exchanges are supported for perpetual and futures markets. For a list of markets supported on your exchange, use the \<exchange\>:markets command. The bot has been extensively tested on these markets, but should work on others as well:

* **Bitmex:**  BTC/USD and ETH/USD
* **Deribit:** BTC-PERPETUAL and ETH-PERPETUAL
* **FTX:** BTC-PERP and ETH-PERP
* **Binance Futures:** BTC/USDT and ETH/USDT

We will add more exchanges based on user demand. Keeping the code cross-compatible over all the exchanges is quite complex so we will only consider adding additional exchanges if a significant number of users request it. Bear in mind that this software is free, so we are under no obligation to add features or provide support, but we will endevour to help out when and where we can. Support can also be found in the [FrostyBot Discord Server](https://discord.gg/yK4U93s). There are a lot of very knowlegeable guys willing to help out if you need assistance.

## Scope      
This bot is specifically designed to execute orders based on webhook alerts received from Tradingview (or from any other system that can send webhooks, such as Discord, Telegram etc). If you have used Autoview in the past you will understand the concept of converting Tradingview strategies and scripts into executable orders on your exchange. However, unlike Autoview, Frostybot makes use of an external web server (VPS, AWS etc), so does not require your PC to be powered on or your browser to be open for it to work.

While is may seem a little complicated to setup and configure for novices, the point of Frostybot is that it's entirely free. You retain total control over your API keys and trading engine. It is perfect for integration projects where you have a trading engine (like a Tradingview strategy or some custom Python script) that you want to integrate with your exchange.

## Requirements
In order to use Frostybot, you will need the following:
* A Linux server which is publicly accessible over the Internet (either by public IP address, DDNS or DNAT). You must be able to access the web server publicly for Frostybot to work. We recommend using the free Ubuntu server available on [Amazon Lightsail](https://lightsail.aws.amazon.com). They also offer a free public IP address. There is also [a handy guide here](https://github.com/CryptoMF/frostybot/blob/master/LIGHTSAIL.md) which will show you how to set up your Lightsail VPS and Frostybot.
* Apache2/Nginx (whichever you prefer). It is **your** responsibility to secure your web server. We accept no responsibility if your server is hacked. We have provided the necessary *.htaccess* file, and it is up to you to ensure that Apache uses it (AllowOveride All setting). We also highly recommend that you secure it with HTTPS (using [LetsEncrypt](https://www.digitalocean.com/community/tutorials/how-to-secure-apache-with-let-s-encrypt-on-ubuntu-18-04) or something similar). It's free, so there's really no excuse for having an insecure web server. If you're using Amazon Lightsail, [here is a guide](https://github.com/CryptoMF/frostybot/blob/master/LIGHTSAILSSL.md) on how to set up a domain name and SSL certificate for your bot.
* PHP 7.2 or higher. You will need the following PHP modules installed (php-curl, php-mbstring, php-json, php-sqlite3, php-cli), and you will need to ensure that PHP is configured on your web server.
* If you want to clone Frostybot from Github, you will need to have git installed.

*Note:* We have a [handy install script](https://github.com/CryptoMF/frostybot/blob/master/INSTALLATION.md#automatic-installation-using-install-script) to automatically install and configure Apache, PHP and Frostybot. Apache and PHP will only be installed if no web server is detected. It will also install git, nano and wget if they are not already installed.

## Installation

* We have a quick and dirty walkthrough on how to [setup Frostybot on Amazon Lightsail](https://github.com/CryptoMF/frostybot/blob/master/LIGHTSAIL.md)
* There is also a tutorial on how to [configure a domain name and SSL for your Lightsail instance]((https://github.com/CryptoMF/frostybot/blob/master/LIGHTSAILSSL.md)
* If you don't want to use Amazon Lightsail, check out our [Installation and Configuration Guide](https://github.com/CryptoMF/frostybot/blob/master/INSTALLATION.md).

## Changelog

The current version of FrostyBot is 0.9. This version constitutes a massive overhaul of the entire codebase and is basically a redevelopment from scratch, with loads of new features and capabilities.

A big shoutout to everybody who has contributed and collaborated on this project. The changelog is available at [FrostyBot Changelog](https://github.com/CryptoMF/frostybot/blob/master/CHANGELOG.md).

## Usage
Its recommended to use sub-accounts to limit risk. First follow the instructions in the [Install Guide](https://github.com/CryptoMF/frostybot/blob/master/INSTALLATION.md). Then configure your Tradingview alerts to call the webhook using the appropriate commands:

* Add your Exchange API account information to the bot configuration using the **config** command as follows:

      ./frostybot config stub=deribitdemo exchange=deribit apiKey="<your api key>" secret="<your api secret>" description="Deribit Test Account" testnet=true

  The **stub** parameter can be anything you like, as long as it's alphanumeric with no spaces. You will use the **stub** when sending commands to the bot, so make it something short and simple. I've just called it "deribitdemo", but if you wanted to use 2 accounts on deribit, you could call the one stub "deribitmain" and the other one "deribittest" for example, it's entirely up to you.

  The **exchange** parameter can be ftx, bitmex or deribit, depending on which exchange you use. The **apiKey** and **secret** are self explanatory, simply add your own api key and secret there (Important: notice the uppercase K in apiKey, it's important to keep it like that). The **description** field can be anything you like, but if you use spaces remember to enclose it in "quotes". Lastly, the **testnet** parameter just lets the bot know which network to connect to. If you want to test it out for a bit, create an api key on the testnet and try it out. The testnet parameter only works for Deribit and Bitmex, because FTX does not have a testnet.

  Lastly, if you are using the FTX exchange, and you are using a subaccount, please also add **subaccount**="\<sub account name\>". This is required by the FTX API, and it will not work until you add it to the config, but only if you're using subaccounts.

* You can list your current config by using the following command:

       ./frostybot config

  If you would like to remove a config, use the command like this:

       ./frostybot config stub=deribitdemo delete=true

  If you just want to update an existing config, just rerun the config command with the same stub name and the other values will be updated.

* Using the deribitdemo config I made above, some example command syntax would be as follows:

      deribitdemo:long size=5000 price=8000 symbol=BTC-PERPETUAL  (This example provides the symbol on the commandline)
      deribitdemo:short size=5000 price=8000 symbol=BTCUSD        (This example uses a symbol mapping in the config to convert BTCUSD to BTC-PERPETUAL)
      deribitdemo:short size=5000 price=8000                      (This example uses the "default" symbol mapping in the config which is mapped to BTC-PERPETUAL)    

*Note:* The order size is always in USD, and the symbol is required for most exchange-specific commands (unless a default symbol mapping has been provided in the config file, in which case that default symbol will be used if no symbol is provided in the command).

To get a list of supported symbols for an exchange, use the following command:

    ./frostybot deribitdemo:markets

* Once all your accounts are configured, ensure that the db/, and log/ directories are writable by the account that you use you run your web server (for example, www-data on Ubuntu). If for some reason you cannot make those directories writable, you can change the location of the directories in the cfg/cfg.config.php file. If you do change those settings, remember to relocate the directories to the new location. Check out the [FrostyBot Install Guide](https://github.com/CryptoMF/frostybot/blob/master/INSTALLATION.md) for more information on how to do this correctly.
* Run some FrostyBot CLI commands to check if you have done the configuration correctly and to ensure that you are able to reach the exchange (Some CLI commands examples are listed further below).
* Check that your bot is accessible over the public internet by browsing to it's address. You should see a message similar to the following:

    {
        "results": {
            "code": 900,
            "message": "Request received from invalid address: 123.69.123.69",
            "type": "ERROR"
        },
        "messages": []
    }

*Note:* This error is expected, but its a good way to check if your bot is responding over http/https. You will receive this specific error because the bot has built-in security that will only accept http/https requests from Tradingview's servers. You can still communicate directly with Frostybot using CLI commands. If you need to communicate to the bot over http, you will need to add your IP address into the "whitelist" by using the following command:

        ./frostybot whitelist add="<ip address>" description="An optional description for your own info"

By default, only Tradingview's servers have access over http/https. Any other machines that you want to allow to access the bot will need to be added to the whitelist. If you want to remove an IP address from the whitelist, use this command:

        ./frostybot whitelist delete="<ip address>"

Note that you cannot delete the default Tradingview addresses in the whitelist, as they are protected against deletion.

Once you've confirmed that Frostybot is responding over the internet, you can start creating Tradingview alerts:

* In Tradingview, create a new alert.
* The Webhook URL should point to your bot address.
* The message box of the Tradingview alert should have the bot command(s), here are some exmaples:

**Tradingview Alert Messagebox Examples:**

    deribit:long symbol=BTC-PERPETUAL size=1000         ($1000 market buy on Deribit BTC-PERPETUAL)
    ftx:short symbol=BTC-PERP size=2000 price=7600      ($2000 limit sell at $7600 on FTX BTC-PERP)
    bitmex:long symbol=BTC/USD size=3x                  (3x long market buy on Bitmex BTC/USD)
    ftx:close size=50% symbol=ETH-PERP                  (Market close 50% of oposition on FTX ETH-PERP)
    deribit:close symbol=ETH-PERPETUAL                  (Close entire ETH-PERPETUAL position on Deribit)
    ftx:short size=200% symbol=BTC-PERP                 (Market sell 200% of your account / 2x short)

*NOTE:* Anything in the message box of the Tradingview alert is interpreted as a command, so do not use the message box for a general description! Also note, you can provide multiple commands in the message box, as long as they are each on a new line. For example, you can close all your open stop loss orders before entering a new position.

Here is an example of a Tradingview alert showing multiple commands:

![TV Alert](https://i.imgur.com/p8YFTah.png)

Once you have found a winning strategy (or made your own using Pinescript on Tradingview), you can set up multiple alerts for long, short, and close (takeprofit/stoploss) orders and let Tradingview and Frostybot fully automate your trading!

*Sample study in Tradingview showing buy, sell and take profit alert points*

![TV Strategy](https://i.imgur.com/isAkusb.png)

*List of tradingview alerts showing long, short and take profit orders*

![List of alert](https://i.imgur.com/qM4OwFJ.png)

**CLI Examples:**

Other than Tradingview firing off webhooks, you can also communicate manually with Frostybot using the commandline interface (CLI). Here are some CLI examples that you can try out.

    ./frostybot deribit:long symbol=BTC-PERPETUAL size=1000     ($1000 market buy on Deribit BTC-PERPETUAL)
    ./frostybot ftx:short symbol=BTC-PERP size=2000 price=7600  ($2000 limit sell at $7600 on FTX BTC-PERP)
    ./frostybot bitmex:long symbol=BTC/USD size=3x              (3x market buy on Bitmex BTC/USD)
    ./frostybot ftx:close size=50% symbol=ETH-PERP              (Market close 50% of position on FTX ETH-PERP)
    ./frostybot deribit:close symbol=ETH-PERPETUAL              (Close entire ETH-PERPETUAL position on Deribit)
    ./frostybot ftx:short size=200% symbol=BTC-PERP             (Market sell 200% of your account / 2x short)
    ./frostybot ftx:markets                                     (Show avaiable markets on FTX exchange)
    ./frostybot bitmex:market symbol=BTC/USD                    (Show market data for BTC/USD on Bitmex)
    ./frostybot ftx:trades                                      (Recent trades, oldest listed first)
    ./frostybot deribit:balance                                 (Current account balance on Deribit)
    ./frostybot bitmex:orders status=open                       (Open orders on Bitmex)
    ./frostybot ftx:cancel id=<orderid>                         (Cancel a specific order on FTX)
    ./frostybot deribit:cancel id=all                           (Cancel all open orders on Deribit)
    ./frostybot bitmex:cancelall                                (Cancel all open orders on Bitmex)
    ./frostybot ftx:cancelall symbol=ADA-PERP                   (Cancel all ADA-PERP orders on FTX)
    ./frostybot bitmex:positions                                (Show current positions on Bitmex, if any)
    ./frostybot deribit:position symbol=BTC-PERPETUAL           (Show current BTC-PERPETUAL position on Deribit)
    ./frostybot ftx:ohlcv symbol=BTC-PERP timeframe=1h          (Show current hourly OHLCV data for FTX BTC-PERP)
    ./frostybot ftx:ohlcv symbol=BTC/USD timeframe=720          (Show current 12hr OHLCV data for Bitmex BTC/USD)
    ./frostybot config                                          (Show your current configuration)
    ./frostybot log                                             (Show log file for bot, last 20 lines by default)

CHANGELOG.md
![logo](https://i.imgur.com/YAME6yT.png "#FrostyBot")

## Changelog

A huge thank you to all of those that have contributed and collaborated on FrostyBot over the past 6 months. A particular shoutout to @Barnz for his huge contribution to the bot. Many things have changed since I started out on this project, and this is still very much a work in progress. I hope to improve on it significantly over the next year.

<table>
  <tr>
    <th>Version</th>
    <th>Changes</th>
  </tr>
  <tr>
    <td>0.1</td>
    <td><ul><li>Initial version</li></ul></td>
  </tr>
  <tr>
    <td>0.2</td>
    <td>
      <ul>
        <li>Code cleanup</li>
        <li>Removed the FLIPLONG and FLIPSHORT commands. Incorporated into LONGENTRY and SHORTENTRY.</li>
        <li>Added POSITION command to show current position</li>
        <li>Added size parameter</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td>0.3</td>
    <td>
      <ul>
        <li>Added ORDERS and CANCEL command to show open orders and cancel orders (limit orders)</li>
        <li>Added price parameter (limit orders)</li>
        <li>Changed output of BALANCE, TRADES and POSITION commands to JSON</li>
        <li>Changed IP whitelist to const to prevent skull fuckery</li>
      </ul>
    </td>
  </tr>
  <tr>
    <td>0.4</td>
    <td>
      <ul>
        <li>Added size parameter support for % of balance (ie. 200pct = 2x)</li>
        <li>Added TAKEPROFIT command with size parameter as % of open order (ie. 50pct)</li>
      </ul>
    </td>  
  </tr>
  <tr>
    <td>0.5</td>
    <td><ul><li>Named changed to Frostybot so I don't get sued</li></ul></td>
  </tr>
  <tr>
    <td>0.6</td>
    <td>
<ul>
<li>ccxt installation via an install.sh script (and cleanup of 750 mb of unneeded files)\n- now bot is usable with testnet or alternate api endpoint urls (base api url is now configurable for exchange)</li>
<li>support bitmexico</li>
<li>flag to turn bot on and off</li>
<li>separate configs per exchange bot, with core functionality in lib extracted so we can have multiple bots without code dupes</li>
<li>allow for space separated command (same syntax as command line) in POST body. This enables:</li>
<li>simpler syntax for users in the alert message box of tradingview</li>
<li>allow pinescript to setup the commands (preconfigure a bot)</li>
<li>let the user override the command by editing the message box instead of a webhook url</li>
<li>webhook url can then just simply point to the bot without querystring which is a bit cleaner</li>
<li>support % besides pct for percentages which is convenient in POST and command line like 'long 50%'</li>
<li>changed api.
<ul>
<li>replaced LONGENTRY, LONGEXIT, SHORTENTRY, SHORTEXIT, TAKEPROFIT</li>
<li>with LONG, SHORT, CLOSE</li>
<li>where CLOSE is replacing LONGEXIT, SHORTEXIT and TAKEPROFIT</li>
<li>CLOSE works the same as LONG / SHORT with percentages or absolute sizes.</li>
</ul>
<li>support for max position (enter multiple longs in a row wont blow you up)</li>
<li>auto adjust order size in case the max limit is reached. Say max position is 100% of your account and you're 50% long. Adding a 'long 100%' would in theory be 150% account size. This last order is auto adjusted to 50% so we end up at the 
maximum position size (which we wanted).</li>
<li>changed api examples to use lower case (easier to type :-))</li>
<li>early exit with more descriptive warnings in case of typo's or wrong parameters.</li>
<li>placed readme and changelog in separate files</li>
<li>added usd balance to log</li>
<li>introduced many bugs, uglified the code :-)</li>
</ul>
</td>
  </tr>
  <tr>
    <td>0.7</td>
    <td>- skipped cause changelog not in sync with version numbers (0.6 was distributed as 0.7, whoops)</td>
  </tr>
  <tr>
    <td>0.8</td>
    <td><ul>
<li>refactor directory/file structure. added lib / config directories and fixed install script for that.</li>
<li>renamed folders and use exchange/pair structure which is more scalable.</li>
<li>currency pairs should now be easy to add in the future.</li>
<li>bot configs are smaller as exchange anomalies are solved in the code.</li>
<li>separated output functions into output.php</li>
<li>this means that the url structure also changes. see INSTALLATION anf README</li>
<li>removed install.sh from the distro. not needed anymore because we include ccxt in the distro for easier setup. licence is included and is MIT, so no issues.</li>
<li>added support for ETH on Deribit and Bitmex</li>
<li>removed hardcoded local ip, added global config for adding your local ip to the whitelist</li>
<li>added exchange configs so you don't have to repeatedly put your keys in for each currency pair</li>
<li>support main accounts in exchange.config.php (default)</li>
<li>can be overridden per bot subaccount in the currency_pair/config.php (recommended)</li>
<li>made main.php more generic. Removed all hardcoded refs to BTC.</li>
<li>added INSTALLATION guide with hopefully easy steps the get going</li>
<li>log file restructuring, placed outside the public www dir (was security issue)</li>
<li>added command 'log' to show the contents of the log file,</li>
<ul>
<li>by default shows last 10 lines</li>
<li>you can give the number of lines as a parameter</li>
</ul>
<li>added 'summary' command. gives an overview for currency pair:</li>
<ul>
<li>balance</li>
<li>position</li>
<li>orders</li>
</ul>
<li>added many comments to make it easier for users to configure the bot</li>
<li>add exchange errors to the log (in raw format)</li>
</ul></td>
  </tr>
  <tr>
    <td>0.9</td>
    <td><ul>
<li>support for FTX exchange added</li>
<li>redeveloped Frostybot from the ground up using object-oriented code as far as possible</li>
<li>all output is now in JSON object format to allow for easy integration and scripting</li>

lib/lib.output.php
<?php

    header('Content-Type: application/json');
    header('Access-Control-Allow-Origin: *');
    set_time_limit(0);
    ignore_user_abort(true);
    ob_start();

    $time_start = microtime(true);

    class outputsObject {

        private $results = [];
        private $messages = [];

        public function addResult($code, $message, $data) {
            $this->results[] = new resultObject($code,$message,$data);
        }

        public function addMessage($code, $type, $message, $data) {
            $this->messages[] = new messageObject($code,$type,$message,$data);
        }

        public function getResults() {
            if (count($this->results) == 1) {
                return $this->results[0]->get();
            } else {
                $output = [];
                foreach($this->results as $result) {
                    $output[] = $result->get();
                }
                return $output;
            }
        }

        public function outputResults() {
            global $time_start;
            $time_end = microtime(true);
            $duration = round($time_end - $time_start,5);
            logger::debug('Script execution time: '.$duration.' seconds');
            $cachestats = $GLOBALS['cachestats']['totals'];
            $cachehits = $cachestats['hit'];
            $cachemisses = $cachestats['miss'];
            $cachetotal = $cachehits + $cachemisses;
            $cachehitratio = ($cachetotal > 0 ? round(($cachehits / $cachetotal) * 100, 0) : 0);
            logger::debug('Cache Stats: '.$cachemisses.' Misses, '.$cachehits.' Hits, '.$cachehitratio.'% Hit Ratio');
            $results = $this->getResults();
            $messages = $this->getMessages();
            $command = $GLOBALS['command'];
            $command = (((is_object($command)) && (isset($command->command))) ? $command->command : $command); 
            if (($results->code == 0) && (!is_object($command)) && (in_array( strtolower($command), ['long', 'short', 'buy', 'sell', 'stoploss', 'takeprofit', 'trailstop', 'close'] ))) {
                notifications::send('order', ['orders' => $results->data, 'balance' => $GLOBALS['balance']]);
            }
            if (($results->code == 0) && (!is_object($command)) && (in_array( strtolower($command), ['cancel', 'cancelall'] ))) {
                notifications::send('cancel', ['orders' => $results->data, 'balance' => $GLOBALS['balance']]);
            }
            $output = (object) [
                'results' => $results,
                'messages' => $messages,
            ];
            echo json_encode($output, JSON_PRETTY_PRINT).PHP_EOL;
            die;
        }

        public function getMessages() {
            $output = [];
            foreach($this->messages as $message) {
                $output[] = $message->get();
            }
            return $output;
        }

    }

    class resultObject {

        public $code;
        public $type;
        public $class;
        public $message;
        public $data;

        public function __construct($code, $message, $data = []) {
            $this->code = $code;
            if (in_array($data,['error'])) {
                $this->type = $data;
            } else {
                $this->type = gettype($data);
                if (is_null($this->type)) {
                    $this->type='none';
                }
                if (strtolower($this->type) == 'object') {
                    $this->class = get_class($data);
                }
                if ((strtolower($this->type) == 'array') && (count($data) > 0)) {
                    $subtype = gettype($data[array_keys($data)[0]]);
                    $allsame = true;
                    foreach ($data as $check) {
                        if (gettype($check) !== $subtype) {
                            $allsame = false;
                        }
                    }
                    if ($allsame === true) {
                        $class = strtolower($subtype) == 'object' ? get_class($data[array_keys($data)[0]]) : $subtype;
                        $this->class = $class;
                    }
                }
                $this->data = $data;
            }
            $this->message = $message;
        }

        public function get() {
            $result = [];
            $result['code'] = $this->code;
            $result['message'] = $this->message;
            if (((is_null($this->type)) || ($this->type == "NULL"))  && (strtolower($this->message) == 'success')) {
                $result['type'] = "SUCCESS";
            } else {
                $result['type'] = $this->type;
            }
            if (!is_null($this->class)) {
                $result['class'] = $this->class;
            }
            if (!is_null($this->data)) {
                $result['data'] = $this->data;
            }
            return (object) $result;
        }

    }

    class messageObject {

        public $code;
        public $type;
        public $message;
        public $data;

        public function __construct($code, $type, $message, $data = []) {
            $this->code = $code;
            $this->type = strtoupper($type);
            $this->message = $message;
            $this->data = $data;
        }

        public function get() {
            $result = [];
            $result['code'] = $this->code;
            $result['type'] = $this->type;
            $result['message'] = $this->message;
            if ($this->data != []) {
                $result['data'] = $this->data;
            }
            return (object) $result;
        }

    }

    $__outputs__ = new outputsObject();

    function outputResult($code,$message,$data = []) {
        global $__outputs__;
        $__outputs__->addResult($code,$message,$data);
        $__outputs__->outputResults();
    }

    function testResult($code,$message,$data) {
        $output = new outputsObject();
        $output->addResult($code,$message,$data);
        return $output->getResults();
    }

    function message($code,$type,$message,$data = []) {
        global $__outputs__;
        $__outputs__->addMessage($code,$type,$message,$data);
    }

    function errorHandler($errno, $errstr, $errfile, $errline) {
        if (!(error_reporting() & $errno)) {
            return false;
        }
        switch ($errno) {
            case E_USER_ERROR:
                $type = 'ERROR';
                break;    
            case E_USER_WARNING:
                $type = 'WARNING';
                break;
            case E_USER_NOTICE:
                $type = 'NOTICE';
                break;
            case E_ERROR:
                $type = 'ERROR';
                break;    
            case E_WARNING:
                $type = 'WARNING';
                break;
            case E_NOTICE:
                $type = 'NOTICE';
                break;
            default:
                $type = 'NOTICE';
                break;
        }
        $logtype = strtolower($type);
        logger::$logtype($errstr.' ('.basename($errfile).':'.$errline.')', $errstr);
        return true;        
    }

    function exceptionHandler($e) {
        $type = strtolower(get_class($e));
        if (!in_array($type,['error','warning','notice','info','debug'])) {
            $message = strtoupper($type).': '.$e->getMessage().' ('.$e->getFile().':'.$e->getLine().')';
            $type='error';    
        } else {
            $message = $e->getMessage().' ('.$e->getFile().':'.$e->getLine().')';
        }
        logger::$type($message, $e->getMessage());
        return true;
    }

    set_error_handler("errorHandler");
    set_exception_handler( "exceptionHandler" );
    ini_set( "display_errors", "off" );
    error_reporting(E_ALL);

    class logger {

        static private function message($type, $message) {
            $message = date('Y-m-d H:i:s').' | '.str_pad(strtoupper($type),7," ",STR_PAD_RIGHT).' | '.$message.PHP_EOL;
            file_put_contents(logfile, $message, FILE_APPEND);
        }

        static public function notice($message) {
            message(0,'NOTICE',$message);
            self::message('notice', $message);
        }

        static public function info($message) {
            message(0,'INFO',$message);
            self::message('info', $message);
        }

        static public function debug($message) {
            if (debug === true) {
                message(1000,'DEBUG',$message);
                self::message('debug', $message);
            }
        }

        static public function warning($message, $notifymsg = null) {
            message(800,'WARNING',$message);
            notifications::send('custom', ['type'=>'Warning', 'message'=>(is_null($notifymsg) ? $message : $notifymsg)]);
            self::message('warning', $message, $notifymsg);
        }

        static public function error($message, $notifymsg = null) {
            //message(900,'ERROR',$message);
            notifications::send('custom', ['type'=>'Error', 'message'=>(is_null($notifymsg) ? $message : $notifymsg)]);
            self::message('error', $message, $notifymsg);
            outputResult(900,$message,'error');
        }

        static public function get($params) {
            $clear = (isset($params['clear']) ? $params['clear'] : false);
            if ($clear == true) {
                file_put_contents(logfile,'');
                logger::notice('Log file cleared');
            }
            $lines = (isset($params['lines']) ? $params['lines'] : 10);
            $filter = (isset($params['filter']) ? $params['filter'] : null);
            $log = explode("\n",file_get_contents(logfile));
            $output = [];
            foreach($log as $key => $entry) {
                if (trim($entry) == "") {
                    unset($log[$key]);
                } else {
                    $logentry = explode(" | ", $entry, 3);
                    if ((is_array($logentry)) && (count($logentry) == 3)) {
                        list($date,$type,$message) = $logentry;
                        if (!is_null($filter)) {
                            if ((strpos(strtolower($message),strtolower($filter)) !== false) || (strpos(strtolower($type),strtolower($filter)) !== false)) {
                                $output[] = (object) [
                                    'datetime' => $date,
                                    'type' => trim($type),
                                    'message' => trim(str_replace("\r","",$message)),
                                ];        
                            }
                        } else {
                            $output[] = (object) [
                                'datetime' => $date,
                                'type' => trim($type),
                                'message' => trim(str_replace("\r","",$message)),
                            ];    
                        }
                    }
                }
            }
            if (!is_null($lines)) {
                $output = array_slice($output,(0-$lines));
            } 
            return $output;
        }


    }

    ob_end_flush();

?>
